src/sign_tx.c

```diff
@@ -0,0 +1,449 @@
+        os_memmove(ctx->buffer, buf, ctx->buffer_len);
+        ctx->current_output++;
+    } else {
+        // end of data we should read. Is there something left on the buffer?
+        if (ctx->buffer_len > 0) {
+            THROW(TX_STATE_ERR);
+        }
+        THROW(TX_STATE_FINISHED);
+    }
+
+    switch (ctx->elem_type) {
+        case ELEM_TOKEN_UID:
+            // not displaying token uid now
+            break;
+        case ELEM_INPUT:
+            // not displaying inputs
+            break;
+        case ELEM_OUTPUT:
+            // check if this is the change output
+            if (ctx->has_change_output && ctx->change_output_index == ctx->decoded_output.index) {
+                if (!verify_change_output(ctx->decoded_output, ctx->change_key_index)) {
+                    THROW(TX_STATE_ERR);
+                }
+            } else {
+                // if it's not change output, raise TX_STATE_READY to display output on screen
+                THROW(TX_STATE_READY);
+            }
+            break;
+    }
+}
+
+tx_decoder_state_e decode_next_element() {
+    volatile tx_decoder_state_e result;
+    BEGIN_TRY {
+        TRY {
+            // read until we reach a displayable element or the end of the buffer
+            for (;;) {
+                _decode_next_element();
+            }
+        }
+        CATCH_OTHER(e) {
+            result = e;
+        }
+        FINALLY {
+        }
+    }
+    END_TRY;
+    PRINTF("decode_next_element result %d\n", result);
+    return result;
+}
+
+/*
+ * Parses the change output info and returns its size. The first byte indicates
+ * whether there's change or not (no change output if byte=0x00). There are 2
+ * possible return values:
+ *   . 1: if there's no change output, it's only 1 byte;
+ *   . 6: if there is, it's [change_exists (1 byte) + output_index (1 byte) + key_index (4 bytes)]
+ */
+static uint8_t parse_change_output_info(uint8_t *in, size_t inlen) {
+    uint8_t *buf = in;
+    assert_length(1, inlen);
+    // check output change. If next byte is greater than 0, there's a change output
+    ctx->has_change_output = (*buf > 0 ? true : false);
+    buf++;
+    if (ctx->has_change_output) {
+        assert_length(5, inlen - 1);
+        ctx->change_output_index = *buf;
+        buf++;
+        ctx->change_key_index = U4BE(buf, 0);
+        buf += 4;
+    }
+    return buf - in;
+}
+
+/*
+ * Prepare the output information that will be displayed.
+ */
+static void prepare_display_output(tx_output_t output) {
+    // first prepare the address + value line
+    unsigned char address[25];
+    pubkey_hash_to_address(output.pubkey_hash, address);
+    uint8_t len = encode_base58(address, 25, ctx->info, sizeof(ctx->info));
+    os_memmove(ctx->info + len, " HTR ", 5);
+    format_value(output.value, ctx->info + len + 5);
+
+    // line1
+    // if there's change output, subtract one
+    uint8_t total_outputs = ctx->has_change_output ? ctx->outputs_len - 1 : ctx->outputs_len;
```

Comment:
What if we have two outputs for ourselves?

Reply:
Only one is considered the change output and only that one is hidden from Ledger display.

Reply:
I don't understand why we are limiting our implementation to only one output. I know that's how our wallet does it today, but it may have future problems.

In the current implementation, what would it show to the user if I decided to send 1,000 HTR to my own address? Both outputs will be mine, so you'd skip the first and show the second? Maybe we should further discuss how to handle these special cases.

Reply:
No, the wallet specifically sends which one is the change output. It'll always have to do it. Right now, we only expect to receive one (or none) change outputs, but we could extend to receive more. I just didn't think it was relevant as we never even discussed having multiple outputs, so it doesn't seem something that will change in the next few months.

Reply:
@msbrogli It's different to ask the wallet to send an amount to your own address and to have a change output.

The change output is automatically generated, the others you decide the amount and the destination address.

We hide on ledger only the change because the user has not selected them on the wallet so it could be confusing.

Reply:
You're right. And I think we should include this explanation in the comments and also update the design.